xds/src/main/java/io/grpc/xds/XdsLoadReportStore.java

```diff
@@ -0,0 +1,140 @@
+  // Cluster level dropped request counts for each category specified in the DropOverload policy.
+  private final ConcurrentMap<String, AtomicLong> dropCounters;
+
+  XdsLoadReportStore(String clusterName) {
+    this(clusterName, new ConcurrentHashMap<Locality, ClientLoadCounter>(),
+        new ConcurrentHashMap<String, AtomicLong>());
+  }
+
+  @VisibleForTesting
+  XdsLoadReportStore(String clusterName,
+      ConcurrentMap<Locality, ClientLoadCounter> localityLoadCounters,
+      ConcurrentMap<String, AtomicLong> dropCounters) {
+    this.clusterName = checkNotNull(clusterName, "clusterName");
+    this.localityLoadCounters = checkNotNull(localityLoadCounters, "localityLoadCounters");
+    this.dropCounters = checkNotNull(dropCounters, "dropCounters");
+  }
+
+  /**
+   * Generates a {@link ClusterStats} containing load stats in locality granularity.
+   */
+  ClusterStats generateLoadReport(Duration interval) {
+    ClusterStats.Builder statsBuilder = ClusterStats.newBuilder().setClusterName(clusterName)
+        .setLoadReportInterval(interval);
+    for (Map.Entry<Locality, XdsClientLoadRecorder.ClientLoadCounter> entry : localityLoadCounters
+        .entrySet()) {
+      XdsClientLoadRecorder.ClientLoadSnapshot snapshot = entry.getValue().snapshot();
+      statsBuilder
+          .addUpstreamLocalityStats(UpstreamLocalityStats.newBuilder()
+              .setLocality(entry.getKey())
+              .setTotalSuccessfulRequests(snapshot.callsSucceed)
+              .setTotalErrorRequests(snapshot.callsFailed)
+              .setTotalRequestsInProgress(snapshot.callsInProgress));
+    }
+    for (Map.Entry<String, AtomicLong> entry : dropCounters.entrySet()) {
+      statsBuilder.addDroppedRequests(DroppedRequests.newBuilder()
+          .setCategory(entry.getKey())
+          .setDroppedCount(entry.getValue().get()));
+    }
+    return statsBuilder.build();
+  }
+
+  /**
+   * Create a {@link ClientLoadCounter} for the provided locality if not present in this {@link
+   * XdsLoadReportStore}. This method needs to be called at locality updates for newly assigned
+   * localities in balancer discovery responses.
+   */
+  void addLocality(Locality locality) {
+    localityLoadCounters.putIfAbsent(locality, new ClientLoadCounter());
+  }
+
+  /**
+   * Discard the {@link ClientLoadCounter} for the provided locality if owned by this {@link
+   * XdsLoadReportStore} to avoid map size growing infinitely. To be called at locality updates when
+   * the provided locality is no longer considered in balancer discovery response.s
+   */
+  void removeLocality(Locality locality) {
+    localityLoadCounters.remove(locality);
+  }
+
+  /**
+   * Intercepts a in-locality PickResult with load recording {@link ClientStreamTracer.Factory}.
+   */
+  PickResult interceptPickResult(PickResult pickResult, Locality locality) {
+    if (!pickResult.getStatus().isOk() || !localityLoadCounters.containsKey(locality)) {
```

Comment:
nit: I would split this into two if-branches, with the second being:
```java
XdsClientLoadRecorder.ClientLoadCounter counter = localityLoadCounters.get(locality);
if (counter == null) {
  return;
}
```

Then you don't call `get()` again below. This has a minor performance advantage that may be negligible. It also helps with readability, because it's very obvious `counter` is not null from this point, but it's less obvious `localityLoadCounters.get(locality)` is not null.

Reply:
ACK.